File: lms/src/main/java/com/lbraz/lms/service/EnrollmentService.java
package com.lbraz.lms.service;

import com.lbraz.lms.dto.EnrollmentRequest;
import com.lbraz.lms.entity.Enrollment;

import java.util.List;
import java.util.UUID;

public interface EnrollmentService extends BaseService<Enrollment, UUID> {
    Enrollment enrollStudent(EnrollmentRequest enrollmentDTO);

    Enrollment completeCourse(UUID enrollmentId);

    void updateExpiredEnrollments(List<UUID> enrollmentIds);
}

File: lms-ui/src/app/services/enrollment.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { BaseService } from './base.service';
import { Enrollment } from '../models/enrollment.model';
import { Observable } from 'rxjs';
import { EnrollmentRequest } from '../models/enrollment-request.model';
import { environment } from '../../environments/environment';

@Injectable({
  providedIn: 'root'
})
export class EnrollmentService extends BaseService<Enrollment, string> {
  private readonly enrollUrl = `${environment.apiUrl}/enrollments/enroll`;
  private readonly completeUrl = (id: string) => `${environment.apiUrl}/enrollments/${id}/complete`;
  private readonly expireUrl = `${environment.apiUrl}/enrollments/expire`;

  constructor(http: HttpClient) {
    super(http, 'enrollments');
  }

  enrollStudent(request: EnrollmentRequest): Observable<Enrollment> {
    return this.http.post<Enrollment>(this.enrollUrl, request);
  }

  completeCourse(enrollmentId: string): Observable<Enrollment> {
    return this.http.patch<Enrollment>(this.completeUrl(enrollmentId), {});
  }

  updateExpiredEnrollments(enrollmentIds: string[]): Observable<void> {
    return this.http.patch<void>(this.expireUrl, enrollmentIds);
  }
}
